### ISSUE TITLE
0.110.1 - Breaks Custom Icons

### ISSUE THREAD

**Issue by josh929800** [Armaell/home-assistant-custom-icons-loader#5]
After upgrading to 0.110.1 the Custom Icons no longer load they were working prior to the upgrade. Note that they did update the way Home Assistant loads Icons in this release. https://www.home-assistant.io/blog/2020/05/20/release-110/

**Comment by shbatm** [comment]
@Armaell -- See this for the new method: https://developers.home-assistant.io/blog/2020/05/09/custom-iconsets/

**Comment by adamoberley** [comment]
Where do we need to put the code for the new method?

**Comment by migo-dev** [comment]
_custom_components/custom_icons/template.js_ line 49/50 is the only time "ha-iconset-svg" is mentioned AFAIK.
Unfortunately I'm not talented enough to implement this myself - but here's how the 'hass bha icons' are now done (including backwards-comp.): https://github.com/hulkhaugen/hass-bha-icons/pull/8/commits

**Comment by adamoberley** [comment]
Looks like you just make a `custom_icons.js` file in your `config/www` folder, add `/local/custom_icons.js` to your Lovelace resources as a Module and then put your icons in the format of the new API in the `custom_icons.js` file you created.
 https://github.com/matt8707/hass-config/commit/40047acc15491ea803878f84245ea5b1d37cb036

**Comment by adamoberley** [comment]
I was able to get my custom icons to work doing this. I had to upload the icons at https://materialdesignicons.com/custom and then use the path from the .xml file that it exports to get them to show correctly in HA.

**Comment by migo-dev** [comment]
Thanks for your help! Got it working :)

**Comment by Nicras** [comment]
For me this doesn't work. The path in the xml is always empty. Seems, my svgs are not in the right format. I would really appreciate if we get an update here

**Comment by Petro31** [comment]
@Nicras clean up your svgs using inkscape and save them as SVG Optimized

**Comment by badguy99** [comment]
I don't think this is going to be a trivial fix for this component. It seems like Fontello uses a different format for the SVG path, than what HA wants now. And it won't draw them using the Fontello format.

**Comment by u8915055** [comment]
Im really trying to figure out the format of the SVG file that can be supported.  I created the custom_icons.js file and tried out the two samples that are in @matt8707 config above, and they work.  So now im trying to take some custom SVG icons that I created and previously incorporated into HA via extra-HTML inclusion, and somehow move them into the custom_icons.js file, and am having no luck.  I read the details referenced above on the API for custom icons, but it doesn't say much about the specific SVG file format.  I gather they have to be 24px x 24px but do they have to be a single path?  Is there support for style attributes?  I have managed to take one of my icons which was previously 1024x1024 px and shrink it to 24x24, and then create a compound path, but there is still a style attribute there of stroke-width and I dont know what to do with that.  Without the style attribute the icon does not work when including the compound path in the custom_icons.js file.  Anyone have an idea on this?  I will fully admit im a beginner.  I am using Inkscape to manipulate these files.  Thanks in advance.

**Comment by fanaticDavid** [comment]
The only custom icons I use are icons from the Philips Hue icon pack. If I open any of their SVG files in Visual Studio Code, this is what I see:

<img width="1117" alt="Schermafbeelding 2020-05-23 om 04 36 09" src="https://user-images.githubusercontent.com/20402931/82719925-011b5600-9caf-11ea-9eb1-052e546d8c4a.png">

The path needed to make this work is what is in the `d` attribute. This is also what the article on the developer blog talks about:

> `path` is the path of the `svg`. This is the string that is in the `d` attribute of the `<path>` element. The `viewBox` is optional and will default to `0 0 24 24`.

This is what my `custom_icons.js` file ends up looking like:

```javascript
const ICON = {
  bulbssultan: '...',
  heroesbloom: '...',
  heroeshueplay: '...',
  heroeslightstrip: '...',
  devicesbridgesv2: '...'
};

async function getIcon(name) {
  return {
    path: ICON[name],
    viewBox: "2 2 28 28"
  };
}

window.customIconsets = window.customIconsets || {};
window.customIconsets['custom'] = getIcon;
```

The `viewBox` is something I needed to define in order to get the size and positioning of the Hue icons right.

**Comment by u8915055** [comment]
Thanks @fanaticDavid .  That's very helpful info. 
I managed to get further with Inkscape.  This is what I have working now.  It allows me to take just about any SVG icon I have and convert it to a path that I can then cut-n-paste into the custom_icons.js file.  Im not using any other online service or anything to clean these up.  

- First off, I open the SVG file in Inkscape.  
- Change the document size to 24x24px.  
- Select all objects in my SVG, group them, and then scale them to something less than 24x24.  Ive been using 23.5 on the longer axis and then scale proportionally.  
- Center the grouped object on the page
- Ungroup object
- This is where it gets tricky. The goal here is to get all objects to be paths.  If you go under Edit->XML Editor you can look and see if any of the objects are represented by anything other than 'Path'.  You might see RECT or CIRCLE that sort of thing.  Each one needs to be converted to a path.  Use the options under the PATH menu to do so.
- Once all objects are paths, you then need to change the fill and stroke of each one to be 'undefined'  Do this by selecting them all and going under Object->Fill and Stroke and set all fill and stroke to be undefined (the ? mark).
- Select all objects again and go to Path->Combine.  Now you should have one path that represents the whole icon.
- Go to Edit->XML Editor and copy the path of the icon there into the appropriate line in custom_icons.js

I hope this makes sense.  Took a lot of screwing around to get this to work.  But it seems to work for me when I want to not go through Fontello or anything like that to incorporate custom icons.  Hope this helps someone.

**Comment by Petro31** [comment]
you can turn objects into paths by joining them together.  My usual process is to:

Ungroup everything (may require multiple ungroups)
Remove all strokes and replace with paths.
select two objects and join them together.

When all said and done, save as SVG optimized.  It will remove all filler, strokes and fills.  Worse case is you open the xml file and remove the styles that are present.

In the end they look like this via text editors:

```
<svg width="24" height="24" version="1.1" viewBox="0 0 24 24">
 <path d="m11 3v10h2v-10h-2zm-3.5605 2.4395c-2.08 1.44-3.4395 3.8405-3.4395 6.5605a8 8 0 0 0 8 8 8 8 0 0 0 5-1.7637v-2.9277a6 6 0 0 1-5 2.6914 6 6 0 0 1-6-6c0-2.17 1.1609-4.0591 2.8809-5.1191l-1.4414-1.4414zm9.1211 0-1.4512 1.4512c1.1685 0.70922 2.0666 1.8063 2.5332 3.1094h2.1035c-0.48459-1.8771-1.6339-3.4863-3.1855-4.5605zm-0.56055 5.5605v2h2v8h2v-10h-4z"/>
</svg>
```

**Comment by badguy99** [comment]
>  
> you can turn objects into paths by joining them together. My usual process is to:
> 
> Ungroup everything (may require multiple ungroups)
> Remove all strokes and replace with paths.
> select two objects and join them together.
> 

If you can show how to do that programmatically in python, I had code that can be put in to the existing code to generate the JS file with new API call in /config/www/custom/custom_icons/ which could then be used by the front end.

For example once processed by Fontello, I get the following for one icon, using the existing component:
```
M800 277c-48 16-119 62-214 158-77 76-137 161-157 213-4 11 6 23 18 21 12-3 29-8 50-18 58-27 111-62 177-128 66-66 101-119 128-177 10-22 15-39 18-51 2-11-9-21-20-18m-111-130c-2-7-9-
11-15-11-2 0-3 0-5 1-43 13-96 48-157 101-72-39-189-86-186-106 6-38-17-59-61-16-43 43-22 66 17 60 20-3 67 116 106 188-47 56-84 112-99 153-3 8 1 17 9 20 8 3 17-1 20-9 20-53 81-136 153-208 105
-106 171-142 208-154 8-2 12-11 10-19m163 211c-60 129-214 284-341 344-52 24-91 29-115 17-3-2-5-4-8-6-32-21-62-45-90-73-177-176-246-392-72-566 171-171 390-104 566 72 28 28 52 58 72 89 1 2 3 4
 4 6 13 23 9 63-16 117
```

However that doesn't work, but the existing SVG before processing by Fontello had the following, which does work with the new API:
```
M25.5963714,18.3262574 C24.0613714,17.8292574 21.8043714,16.3522574 18.7423714,13.2892574 C16.2863714,10.8352574 14.3593714,8.12425745 13.7353714,6.45925745 C13.6003714,6.09925745
13.9233714,5.73225745 14.3003714,5.80525745 C14.6873714,5.88025745 15.2163714,6.04225745 15.9033714,6.36225745 C17.7513714,7.22225745 19.4493714,8.34025745 21.5703714,10.4612574 C23.6913714
,12.5822574 24.8083714,14.2792574 25.6683714,16.1282574 C25.9953714,16.8292574 26.1583714,17.3652574 26.2303714,17.7552574 C26.2983714,18.1232574 25.9523714,18.4412574 25.5963714,18.3262574
 M22.0383714,22.5102574 C21.9703714,22.7212574 21.7743714,22.8552574 21.5633714,22.8552574 C21.5123714,22.8552574 21.4603714,22.8472574 21.4093714,22.8312574 C20.0353714,22.3872574 18.32437
14,21.2742574 16.3833714,19.5782574 C14.0863714,20.8372574 10.3313714,22.3272574 10.4283714,22.9692574 C10.6143714,24.1962574 9.87837144,24.8752574 8.49237144,23.4892574 C7.10737144,22.1032
574 7.78537144,21.3682574 9.01437144,21.5542574 C9.65837144,21.6512574 11.1553714,17.8472574 12.4203714,15.5552574 C10.9113714,13.7592574 9.71737144,11.9542574 9.23437144,10.6642574 C9.1363
7144,10.4052574 9.26837144,10.1172574 9.52637144,10.0202574 C9.78237144,9.92025745 10.0723714,10.0542574 10.1693714,10.3122574 C10.8013714,11.9972574 12.7673714,14.6712574 15.0623714,16.966
2574 C18.4343714,20.3382574 20.5423714,21.4992574 21.7173714,21.8802574 C21.9803714,21.9652574 22.1233714,22.2472574 22.0383714,22.5102574 M27.2473714,15.7312574 C25.3283714,11.6092574 20.4
093714,6.65425745 16.3353714,4.73925745 C14.6743714,3.95925745 13.4353714,3.82225745 12.6833714,4.20725745 C12.5903714,4.26525745 12.4993714,4.33025745 12.4083714,4.39025745 C11.4013714,5.0
5525745 10.4263714,5.82625745 9.52137144,6.73025745 C3.88037144,12.3602574 1.65337144,19.2642574 7.21737144,24.8192574 C12.7083714,30.3022574 19.6983714,28.1502574 25.3373714,22.5192574 C26
.2293714,21.6302574 26.9913714,20.6742574 27.6493714,19.6852574 C27.6903714,19.6242574 27.7323714,19.5632574 27.7723714,19.5022574 C28.1853714,18.7422574 28.0513714,17.4582574 27.2473714,15
.7312574
```

**Comment by adamoberley** [comment]
Yep, like others have said, you need one single path with no styles. I used Illustrator so I outlined all my strokes and combined all my layers into 1 compound path and exported with the 24x24 artboard for size. After that I could put them in the mdi custom icon exporter to get the path.

**Comment by Nicras** [comment]
Thanks for all these walkthroughs. Can anybody tell me, what fontello did to these svgs, that made them usable? It would be much easier to let them do the work and only extract the path from the correct place. I'm talking about hundreds of icons and I don't want to rework them all this way...

**Comment by shbatm** [comment]
> M800 277c-48 16-119 62-214 158-77 76-137 161-157 213-4 11 6 23 18 21 12-3 29-8 50-18 58-27 111-62 177-128 66-66 101-119 128-177 10-22 15-39 18-51 2-11-9-21-20-18m-111-130c-2-7-9-
> 11-15-11-2 0-3 0-5 1-43 13-96 48-157 101-72-39-189-86-186-106 6-38-17-59-61-16-43 43-22 66 17 60 20-3 67 116 106 188-47 56-84 112-99 153-3 8 1 17 9 20 8 3 17-1 20-9 20-53 81-136 153-208 105
> -106 171-142 208-154 8-2 12-11 10-19m163 211c-60 129-214 284-341 344-52 24-91 29-115 17-3-2-5-4-8-6-32-21-62-45-90-73-177-176-246-392-72-566 171-171 390-104 566 72 28 28 52 58 72 89 1 2 3 4
>  4 6 13 23 9 63-16 117

Is this straight from Fontello or processed by something? Looks like you're missing spaces between the moves (`63-17` should be `63 -17`).

If straight from Fontello, have you tried https://icomoon.io/

**Comment by badguy99** [comment]
> Is this straight from Fontello or processed by something? Looks like you're missing spaces between the moves (`63-17` should be `63 -17`)

This is what is in the `/config/custom_components/custom_icons/cache/custom.<epoch>.js` file the component creates, but that is a straight copy from what Fontello produces using the instructions for the component.

**Comment by shbatm** [comment]
I confirmed on one of my icons, adding the spaces before the negatives, and around the alphabetic commands (`c` `m` and `z`) fixed the issue I was having with them loading.

**Comment by u8915055** [comment]
I have my custom_icons.js file populated now with about 15-20 icons.. the icons all load, however when I load my pages in home assistant that have the icons, I find that maybe 1/2 of them will populate 90% of the time.. if I reload the page, some times the others will load, and if I reload again, they all may load.. So they all work, just not sure why they dont all load every time.  Anyone else finding this?

**Comment by adamoberley** [comment]
I’ve noticed similar behavior from one of my icons. I’ve only noticed it through a browser and emptying the cache has fixed it each time.

**Comment by badguy99** [comment]
The only thing I've had it not load is the custom icons I use on the top menu bar. I need to do some debugging, but I think this is probably a HA bug, as when I open the edit screen it loads the icon correctly.

**Comment by badguy99** [comment]
As an update. To get my custom icons in the top views menu bar to load correctly, I had to add the following in my configuration yaml rather than it being added as a resource for Lovelace via the GUI
```yaml
frontend:
  extra_module_url: /hacsfiles/custom_icons/custom_icons.js
```

**Comment by u8915055** [comment]
Hi @ adamoberley, I found that clearing the cache was only working for short time and then the problem would return.  However I seemed to have found a solution.   In seeing what @badguy99 wrote I tried that.. However when I moved the reference from resources to front end, the icons stopped loading all together.  So I moved it back to resources but moved it to the top of my resource definitions.  Not sure if it was the order in the resource definitions or whether it was removing it and adding it back, but it works like a champ now.  Even with using the custom header integration with these icons.

**Comment by NovaGL** [comment]
Use the instructions on this site to create your own custom icon file.
Works great!

PS: It's similar to what's written above, not sure which one is better.

https://community.home-assistant.io/t/lovelace-collasping-ps4-controller-card-with-icons/198040

**Comment by adamoberley** [comment]
@NovaGL Glad to help. Ironically that post with the PS4 logos is also mine, so if you need any help with either, let me know!

**Comment by NovaGL** [comment]
Haha. The way they tackled the problem was different so I didn't realise.

I think the way you outlines here is better but the other one us easier to understand.

Is there a need for a viewbox as mentioned above?
Also I'm guessing if I want a different prefix besides 'custom' I just change the last line.

For those who are struggling the way i get my icons to work is I make sure there is only one path. E.g. Compound Path in illustrator.

Then I minify it using this site:
https://jakearchibald.github.io/svgomg/

**Comment by mdezzi** [comment]
@NovaGL I had to use the custom viewbox to "resize" the Hue icons to get them to show correctly. I havent had a chance to try any other icon sources. 

Has anyone figured out a way to reload the java resources in Fully Kiosk? I have a Amazon Fire 7 that WILL NOT load the custom icons. I have cleared cache, rebooted, etc.

**Comment by kongo09** [comment]
I've written a walkthrough how I solved it only with Inkscape plus one Inkscape extension here:
https://community.home-assistant.io/t/implement-some-custom-icons-ive-created/13884/52?u=kongo09
Thank you @u8915055 for clarifying most of it for me.
